File: tornado/server-05-db-translate.py
#!/usr/bin/env python
import tornado.ioloop
import tornado.web
import asyncmongo

from utils import validate_params

MONGO_HOST = 'localhost'
MONGO_PORT = 27017
MONGO_DB = 'pycon'

PORT = 8080

class MainHandler(tornado.web.RequestHandler):

    @property
    def db(self):
        if not hasattr(self, '_db'):
            self._db = asyncmongo.Client(dbname=MONGO_DB, host=MONGO_HOST, port=MONGO_PORT, 
                                            maxcached=10, maxconnections=50, pool_id='mypool')
        return self._db

    @tornado.web.asynchronous
    def get(self):
        if not validate_params(self, ('data','user')): return

        def translation_callback(translator, output_str):
            if translator.error:
                self.set_status(500)
                self.write("Internal Server Error. %s" % translator.error)
            else:
                self.write(output_str)
            self.finish()

        def getuser_callback(user, error):
            if error:
                self.set_status(500)
                self.write("Internal Server Error. %s" % error)
                self.finish()
            elif not user:
                self.set_status(403)
                self.write("Forbidden.")
                self.finish()
            else:
                # Connect to translator service
                from tortranslator import TorTranslator
                trans = TorTranslator('localhost', 8010)
                import functools
                trans.translate(self.get_argument('data'), functools.partial(translation_callback, trans))

        self.db.users.find_one({'name': self.get_argument('user')}, callback=getuser_callback)

application = tornado.web.Application([
    (r"/translate/", MainHandler),
])

if __name__ == "__main__":
    application.listen(PORT)
    print "Listening on http://localhost:%s/translate/ params:name,user" % PORT
    tornado.ioloop.IOLoop.instance().start()

File: nodejs/server-05-db-translate.js
var http = require("http");
var url = require('url');
var net = require('net');
var translator = require('./ndtranslator');
var Db = require('mongodb').Db;
var Server = require('mongodb').Server;

var SVC_HOST = 'localhost';
var SVC_PORT = 8010;
var DB_HOST = 'localhost';
var DB_PORT = 27017;

var PORT = 8080

http.createServer(function(request, response) {
    if (request.url == '/favicon.ico'){
        response.statusCode = 404;
        response.end();
        return;
    }
    var parts = url.parse(request.url, true);
    
    // Connect to mongo
    var server = new Server(DB_HOST, DB_PORT, { auto_reconnect: false });
    var db = new Db('pycon', server, {safe: true});
    db.open(function(err, db) { 
        db.collection('users', function(err, users_col) {
            users_col.findOne({name: parts.query.user}, function(err, user) {
                if (user == null){
                    response.statusCode = 403;
                    response.write("Insufficient Funds");
                    response.end();
                } else {
                    // We have a valid user, so do the translation
                    doTranslate();
                }
                db.close();
            });
        });
    });

    function doTranslate() {
        translator.translate(SVC_HOST, SVC_PORT, parts.query.data, function(err, output_str){
            if (err == null) {
                response.write(output_str);
            } else {
                response.statusCode = 500;
                response.write("Internal Server Error: " + err);
            }
            response.end();
        });
    }

}).listen(PORT);

console.log("Listening on http://localhost:" + PORT + " params:data,user");
